### ISSUE TITLE
[CityTiler] 3D Tiles with textures are almost always bigger than the input CityGML

### ISSUE THREAD

**Issue by jailln** [VCityTeam/py3dtilers#66]
I've observed that in most cases, the 3D Tiles created is bigger than the input CityGML by a factor of 2 of 3 (only happens for conversions with textures). It seems odd since the CityGML is based on XML and not optimized while 3D Tiles is mainly binary and a bit of json and is meant to be optimized. Sometimes though, the 3D Tiles produced is smaller than the input CityGML but I have not been able to find out why yet. I think this issue may be related to #65 ? but it needs to be investigated more thoroughly.

**Comment by clementcolin** [comment]
Quick question: do you compare 3D Tiles created with the CityGML and the textures associated, or only the CityGML files?

If it is the former, I agree that's an odd behavior and results and needs to be investigated more thoroughly.

**Comment by jailln** [comment]
Yes I compare the 3D Tiles and the textures with the CityGML and the textures :slightly_smiling_face:

**Comment by clementcolin** [comment]
I think that this happens because, when we are grouping the textures in atlas, it is hard to find the proper size of the resulting atlas given a set of textures, we approximate it but not perfectly, thus we create non-necessary information (the black pixels) that are not in the original textures.

Do you have, by any chance, a clue to find the optimized size of an atlas, given a set of textures ? We follow this [algorithm](https://blackpawn.com/texts/lightmaps/) to create the atlas, yet, it does not talk about this problem.

**Comment by jailln** [comment]
> I think that this happens because, when we are grouping the textures in atlas, it is hard to find the proper size of the resulting atlas given a set of textures, we approximate it but not perfectly, thus we create non-necessary information (the black pixels) that are not in the original textures.

This seems like a good lead indeed!

> Do you have, by any chance, a clue to find the optimized size of an atlas, given a set of textures ? We follow this [algorithm](https://blackpawn.com/texts/lightmaps/) to create the atlas, yet, it does not talk about this problem.

No I don't and by looking over the internet it seems that it depends on the algorithms and that many algorithm start with a fixed height and unlimited width, try to place the textures, then update the width and height and place the texture again, and so on, until finding the best width/height combination. The author of the blog post you used left is e-mail at the end of the page for contacts and questions though, maybe it's worth contacting him ?

An other solution would be to find the name of this algorithm (guillotine texture packing algorithm ? shelf texture packing algorithm ? etc.) and then to find how to deal with the size for this algorithm ?

Some interesting resources that I came across:
* https://www.codeproject.com/Articles/210979/Fast-optimizing-rectangle-packing-algorithm-for-bu
* https://mozillagfx.wordpress.com/2021/02/04/improving-texture-atlas-allocation-in-webrender/
* http://pds25.egloos.com/pds/201504/21/98/RectangleBinPack.pdf

**Comment by LorenzoMarnat** [comment]
See [comment](https://github.com/VCityTeam/py3dtilers/issues/73#issuecomment-1131489654) about [greedypacker](https://github.com/solomon-b/greedypacker).
